easybuild/easyconfigs/p/PyTorch/PyTorch-1.10.0_skip_cmake_rpath.patch

```diff
@@ -0,0 +1,195 @@
+# explicitely by defining INSTALL_RPATH, but only adding directories that do not match to the "stubs" regex
+# It has been upstreamed in this PR https://github.com/pytorch/pytorch/pull/68912 (not accepted yet at the time of writing)
+diff -Nru pytorch.orig/caffe2/CMakeLists.txt pytorch/caffe2/CMakeLists.txt
+--- pytorch.orig/caffe2/CMakeLists.txt	2021-11-17 11:46:01.797337624 +0100
++++ pytorch/caffe2/CMakeLists.txt	2021-11-18 19:05:35.637707235 +0100
+@@ -630,8 +630,33 @@
+     else()
+       set(DELAY_LOAD_FLAGS "")
+     endif()
+-    target_link_libraries(caffe2_nvrtc ${CUDA_NVRTC} ${CUDA_CUDA_LIB} ${CUDA_NVRTC_LIB} ${DELAY_LOAD_FLAGS})
++    # message("CUDA_NVRTC: ${CUDA_NVRTC}")
++    # message("CUDA_NVRTC_LIB: ${CUDA_NVRTC_LIB}")
++    # message("CUDA_CUDA_LIB: ${CUDA_CUDA_LIB}")
++    # message("DELAY_LOAD_FLAGS: ${DELAY_LOAD_FLAGS}")
++    # if(CUDA_CUDA_LIB MATCHES "stubs")
++    #   message("stubs libraries found in the CUDA_CUDA_LIB: ${CUDA_CUDA_LIB}")
++    # else()
++    #   message("Stubs libs not found in CUDA_CUDA_LIB: ${CUDA_CUDA_LIB}")
++    # endif()
++    # Make sure the CUDA stubs folder doesn't end up in the RPATH of CAFFE2_NVRTC:
++    set(CAFFE2_NVRTC_LIBS ${CUDA_NVRTC} ${CUDA_CUDA_LIB} ${CUDA_NVRTC_LIB})
++    foreach(LIB IN LISTS CAFFE2_NVRTC_LIBS)
++      message("LIB: ${LIB}")
++      if(LIB MATCHES "stubs")
++        message("Filtering ${LIB} from being set in caffe2_nvrtc's RPATH, because it appears to point to the CUDA stubs directory, which should not be RPATHed.")
++      else()
++        cmake_path(GET LIB PARENT_PATH LIB_PATH)
++        message("LIBPATH: ${LIB_PATH}")
++	list(APPEND CAFFE2_NVRTC_RPATH ${LIB_PATH})
++      endif()
++    endforeach()
++    message("CAFFE2_NVRTC_RPATH: ${CAFFE2_NVRTC_RPATH}")
++    set_target_properties(caffe2_nvrtc PROPERTIES INSTALL_RPATH_USE_LINK_PATH FALSE)
++    set_target_properties(caffe2_nvrtc PROPERTIES INSTALL_RPATH "${CAFFE2_NVRTC_RPATH}")
++    target_link_libraries(caffe2_nvrtc ${CAFFE2_NVRTC_LIBS} ${DELAY_LOAD_FLAGS})
+     target_include_directories(caffe2_nvrtc PRIVATE ${CUDA_INCLUDE_DIRS})
++#    message(FATAL_ERROR "STOP HERE, we're debugging")
+     install(TARGETS caffe2_nvrtc DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+     if(USE_NCCL AND BUILD_SPLIT_CUDA)
+       list(APPEND Caffe2_GPU_SRCS_CPP
+diff -Nru pytorch.orig/test/cpp/api/CMakeLists.txt pytorch/test/cpp/api/CMakeLists.txt
+--- pytorch.orig/test/cpp/api/CMakeLists.txt	2021-11-17 11:46:02.991350652 +0100
++++ pytorch/test/cpp/api/CMakeLists.txt	2021-11-18 19:06:41.207423777 +0100
+@@ -61,6 +61,22 @@
+     ${CUDA_CUDA_LIB}
+     ${TORCH_CUDA_LIBRARIES})
+ 
++  # Make sure the CUDA stubs folder doesn't end up in the RPATH of test_api:
++  set(TEST_API_LIBS ${CUDA_LIBRARIES} ${CUDA_NVRTC_LIB} ${CUDA_CUDA_LIB} ${TORCH_CUDA_LIBRARIES})
++  foreach(LIB IN LISTS TEST_API_LIBS)
++    message("LIB: ${LIB}")
++    if(LIB MATCHES "stubs")
```

Comment:
@casparvl I wonder if we should make this a bit stricter?

In theory, this is going to be major trouble if `stubs` is somewhere else in the full path to libraries.

Does `MATCHES` support a regular expression?
Can we only match if either `CUDA` or `cuda` also matches in the path to the library?

Reply:
Well, I'm no `CMake` expert, but I think we could do
```
if((LIB MATCHES "stubs") AND ( LIB MATCHES "cuda" OR (LIB MATCHES "CUDA")))
...
```
Not sure if we should be too worried here though: the `TEST_API_LIBS` pretty much only refer to CUDA based libraries. And the worst thing that can happen on an unintended match is that something is _not_ `RPATH`-ed by `CMake`.

The reverse (something being `RPATH`-ed that turns out to be a real 'stubs' library, i.e. one that doesn't provide real implementations) is pretty much _always_ bad. Even if it's called `stubs` and is _not_ a `CUDA` library, it most likely has no business being `RPATH`-ed, because it probably contains a stub implementation of something else.

I'd not take such a 'generic' approach for our `RPATH` wrappers in general. But since this is a specific patch for `PyTorch`, and since it only affects the libraries in `TEST_API_LIBS`, it's already pretty specific in my opinion.

Reply:
Also, note that this is already a _way_ more targetted solution than the PyTorch people themselves came up with, which was basically: stop `RPATH`-ing anything :)